deltametrics/section.py

```diff
@@ -1,10 +1,180 @@
+import abc
+
 import numpy as np
-from scipy import stats
+from scipy import stats, sparse
+
 import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+
+from . import cube
+from . import plot
+
+
+class SectionVariableMA(np.ma.MaskedArray):
+    """Section variable.
+
+    We subclass the numpy MaskedArray, in order to add `ndarray` methods to
+    the object.
+
+    I think this is the preferred way of handling the section variable, but I
+    need to get it implemented correctly to use the numpy C libraries still.
+
+    .. warning:: NOT IMPLEMENTED
+    """
+
+    def __new__(cls, _arr, _preserved,
+                _preserved_voxel_count, _preserved_voxel_eta,
+                _preserved_height):
+        """Construct the array from _arr.
+
+        In this section we validate the inputs, and then apply the mask
+        """
+        obj = np.asarray(_arr).view(cls)
+        obj._mask = ~_preserved
+        # obj._is_spacetime = True
+        obj._preserved_voxel_count = _preserved_voxel_count
+        obj._preserved_voxel_eta = _preserved_voxel_eta
+        obj._preserved_height = int(_preserved_height)
+        nx = obj._data.shape[1]
+        obj._strata = np.zeros(shape=(obj._preserved_height + 1, nx))
+        for j in np.arange(nx):
+            obj._strata[0:obj._preserved_voxel_count[j], j] = obj._data[
+                0:obj._preserved_voxel_count[j], j]
+        return obj
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self._mask = getattr(obj, '_mask', None)
+        self._preserved_voxel_count = getattr(
+            obj, '_preserved_voxel_count', None)
+        self._preserved_height = getattr(obj, '_preserved_height', None)
+        self._preserved_voxel_eta = getattr(obj, '_preserved_voxel_eta', None)
+        super().__array_finalize__(obj)
+
+    def __getitem__(self, item):
+        return self._data[item]
+
+    def as_spacetime(self):
+        return self._data
+
+    def as_strata(self):
+        return self._strata
+
+
+class SectionVariable(object):
+    """Section variable.
+
+        This could be one way to implement functionality of section
+        variables. We would have to implement all the functionality of the
+        ndarray ourselves. It's more readable to do it this way, but much
```

Comment:
Why would we have to implement everything from `ndarray` ourselves? This isn't clear to me.

Reply:
If we create a new object, it doesn't have a, e.g., `__add__` method by default. 

Alternatively, we can *subclass* the ndarray, and then we get all of its methods. I want to do this, but I haven't quite figured it out.

Reply:
Ok, I think I just don't understand the non-inheritance here. Don't worry about it, I'll learn more later, and as it stands, I think this looks good!

Reply:
I am now subclassing the MaskedArray for all of the `SectionVariable` types, so they will all support arbitrary math and slicing :+1: